ui/components/app/signature-request/signature-request.component.js

```diff
@@ -45,9 +45,72 @@ export default class SignatureRequest extends PureComponent {
       hardwareWalletRequiresConnection,
     } = this.props;
     const { address: fromAddress } = fromAccount;
-    const { message, domain = {} } = JSON.parse(data);
+    const { message, domain = {}, types, primaryType } = JSON.parse(data);
     const { metricsEvent } = this.context;
 
+    const solidityTypes = [
+      'bool',
+      'int',
+      'fixed',
+      'address',
+      'bytes',
+      'string',
+    ];
+    const sanitizeMessage = (msg, baseType) => {
+      let baseTypeType;
+      if (types) {
+        baseTypeType = types[baseType];
+        if (!baseTypeType) {
+          return msg;
+        }
+      } else {
+        return msg;
+      }
+
+      const sanitizedMessage = {};
+      const msgKeys = Object.keys(msg);
+      msgKeys.forEach((msgKey) => {
+        const definedType = Object.values(baseTypeType).find(
+          (ptt) => ptt.name === msgKey,
```

Comment:
What does `ptt` mean? We should replace this with a better variable name. e.g. `property`